### PR TITLE
Boolean Operands cause ChangeNestedForeachIfsToEarlyContinueRector to…

### PR DESCRIPTION
**Pull request by derrickschoen**
… produce code that gives a different output from the original code.

### CODE REVIEW

**Comment by TomasVotruba** [comment]
Do you have an idea how to fix it?

**Comment by derrickschoen** [comment]
> Do you have an idea how to fix `ChangeNestedForeachIfsToEarlyContinueRector`?

This codebase is way over my head.

My best guess would be to simplify what the rector tries to do.
Make it so that it only tries to wrap the if logic in parenthesis, then negate it, then early continue.

Ex
## Minimal PHP Code Causing Issue
```
<?php
$loaded = true;
$usePackages = true;
$notBad = true;
$partPackageList = [ new \StdClass(), null, 'string', 0, [] ];
$payload = [];

foreach( $partPackageList as $partPackage )
{
    if( ! is_null( $partPackage ) && ! is_object($partPackage) )
    {
        if ($loaded || ($usePackages && $notBad))
        {
            $payload[] = $partPackage;
        }
    }
}
```

## Expected Behaviour
```
<?php
$loaded = true;
$usePackages = true;
$notBad = true;
$partPackageList = [ new \StdClass(), null, 'string', 0, [] ];
$payload = [];

foreach( $partPackageList as $partPackage )
{
	if (! (
           ! is_null( $partPackage ) && ! is_object($partPackage))
        ) {
		continue;
	}
	if (! (
            $loaded || ($usePackages && $notBad))
        ) {
		continue;
	}
	$payload[] = $partPackage;
}
```

**Comment by TomasVotruba** [comment]
Related demo link: https://getrector.org/demo/88a502a5-5c37-4c2c-af8b-63c137cdfb4d

File: rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_bool_oprnds.php.inc [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            if(! is_null( $partPackage ) && ! is_object($partPackage) )
+            {
+                if ($loaded || ($usePackages && $notBad))
+                {
+                    $payload[] = $partPackage;
+                }
+            }
+        }
+
+        return json_encode($payload);
```
Comment (TomasVotruba) [inline rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_bool_oprnds.php.inc:28]: The snippet is quite large to cover with test only the problematic spot. 

2 variables and ~5 lines of code in method would be ideal.

<br>

What lines can be remove to still show the bug?

Comment (derrickschoen) [reply rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_bool_oprnds.php.inc:25]: Is it better now?

Comment (TomasVotruba) [reply rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_bool_oprnds.php.inc:25]: Better. Is that the least code needed?

E.g.?

```diff
-$partPackageList = [ new \StdClass(), null, 'string', 0, [] ];
+$partPackageList = [];
```

Comment (derrickschoen) [reply rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_bool_oprnds.php.inc:25]: I've stripped it down as much as I can while reproducing the bug.

Comment (TomasVotruba) [reply rules/solid/tests/Rector/Foreach_/ChangeNestedForeachIfsToEarlyContinueRector/Fixture/multi_exprs_with_bool_oprnds.php.inc:25]: Great, thanks!

**Comment by TomasVotruba** [comment]
Merged locally. It was missing `continue;` statements :laughing: 

Thanks for the test case!

It might need more coverage with `==`, as they're impossible to inverse, e.g. this is not correct inversion: 

```diff
-1 != '1'
+1 == '1'
```

But provided test now passes.

**Comment by derrickschoen** [comment]
> Merged locally. It was missing `continue;` statements 😆
> 
> Thanks for the test case!
> 
> It might need more coverage with `==`, as they're impossible to inverse, e.g. this is not correct inversion:
> 
> ```diff
> -1 != '1'
> +1 == '1'
> ```
> 
> But provided test now passes.

I messed up the bug replication when I removed the variables.
I made a new commit to my branch, not sure if I open a new PR or what.

There is a bug where if there is a nested if with an "OR" and all conditions evaluate to true, the rector changes the output.

**Comment by TomasVotruba** [comment]
> I made a new commit to my branch, not sure if I open a new PR or what.

The branch is merged now. Open new PR

**Pull request merged**